Ref counted ptr for modern cpp

rpatters1 opened the pull request:
This is a re-implementation of `RefCountedPtr` as a wrapper of `std::shared_ptr` for your consideration. I did some unit testing, and it seems to work. I don't know how much other support it needs before it could be merged, so I am posting it as a Draft for comment at this point.

The only thing I was a little uncertain of was this:

```c++

    RefCountedPtr<T>& operator=(T* const p)
    {
        if (p != m_p.get())
        {
            RefCountedPtr<T> tmp(p);
            std::swap(m_p, tmp.m_p);
        }

        return *this;
    }
```

The rest seems obvious.

I've run my full regression test on it, and it did not cause any runtime errors, for what that's worth.

kunitoki commented:
How do you recover your `RefCountedPtr` from the lua stack or a `LuaRef` ?

rpatters1 commented:
The same way the other one did? I'm guessing the `ContainerTraits` at the bottom is the magic. I just stripped out the crap in the old one and replaced it with a `shared_ptr`.

kunitoki commented:
```
    using TestC = TestClass<int>;

    luabridge::getGlobalNamespace(L)
        .beginNamespace("ns")
            .beginClass<TestC>("TestClass")
                .addFunction("getValue", &TestC::getValue)
            .endClass()
        .endNamespace();

    {
        luabridge::RefCountedPtr<TestC> c(new TestC(42));
        auto result = luabridge::setGlobal(L, c, "c");
        ASSERT_TRUE(result);
    }

    runLua("result = c");
    ASSERT_TRUE(result().isInstance<luabridge::RefCountedPtr<TestC>>());
    auto x = result<luabridge::RefCountedPtr<TestC>>();
```

this fails to compile

kunitoki commented:
Is this enough for your use case ? I don't see the need to have a half broken `RefCountedPtr` when we could use `std::shared_ptr` instead

```cpp

    using TestC = TestClass<int>;

    luabridge::getGlobalNamespace(L)
        .beginNamespace("ns")
            .beginClass<TestC>("TestClass")
                .addFunction("getValue", &TestC::getValue)
            .endClass()
        .endNamespace();

    {
        std::shared_ptr<TestC> a = std::make_shared<TestC>(42);
        auto result = luabridge::setGlobal(L, a, "a");
        ASSERT_TRUE(result);
    }

    runLua("result = a");
    ASSERT_TRUE(result().isInstance<TestC*>());
    auto x = result<TestC*>();
    EXPECT_EQ(42, x->getValue());
```

As soon as you want to retrieve the reference count out of lua (while not inheriting from enable_shared_from_this) you'll get a compilation error:

```cpp

LuaBridge3/Source/LuaBridge/detail/TypeTraits.h:91:13 Static_assert failed due to requirement 'detail::dependent_false<(anonymous namespace)::TestClass<int>>' "Failed reconstructing the reference count of the object instance, class must inherit from std::enable_shared_from_this"
```

kunitoki commented:
What i'm saying is that this is extremely dangerous and should be avoided.

If lua is the only one keeping a reference to the underlying shared_ptr, you need to make sure there is no side effect garbage collection run between you get the pointer to your object out of lua and you use it in C++. Sometimes this is hard to enforce, even with static/lifetime analysis.

Consider this snippet:

```cpp

std::shared_ptr<TestC> a = std::make_shared<TestC>(42);
(void) luabridge::setGlobal(L, a, "a");
a.reset();

runLua("result = a; a = nil");
auto x = result<TestC*>();

// ... here some code callback triggers, and someone calls `collectgarbage`
// at this point lua recycled the expired refcounts and deleted the last instance of `a` !

x->getValue(); // boom !!
```

I'm not 100% fine allowing this, but can bend my will if someone else want to shot himself in the foot. Should not be the preferred way the library suggest to work.

rpatters1 commented:
IRL has been literally all day long, so I'm tardy in replying. It's quite clear you don't want to implement any kind of non-intrusive pointer, I am beginning to understand your reasoning. Let me explain where I'm coming from:

I am working with a C++ framework written about 20 years ago. It would be more accurate to call it C-with-classes. It has been exported into Lua with LuaBridge for nearly as long, and there are probably hundreds of thousands of line of Lua code written for it that I don't want to break.

One of the recurring design patterns is to implement `Create` functions that create instances and send them back to the caller. They are basically a flavor of factory. The dangerous thing about them for C/C++ callers is that the caller has to destroy the objects.

But for Lua callers, the cleanup is handled for them. When I inherited the framework, all such created instances were saved in a pool and deleted when the Lua session ended. This has many drawbacks that I don't need to list.

For all its flaws, `RefCounterPtr` is a much better solution because:

- Multiple Lua sessions can now run simultaneously (a feature I have added).
- Lua gc can delete the object when they go out of scope instead of everything being deleted after the Lua state was closed.

But if I'm honest, the best solution would be to offer an easy way for my code to create it on the Lua stack directly. (Maybe this already exists?)

The pattern right now is:

```c++
/** \brief Create object with new and C++ caller has to delete it themselves. */
MyObject * CreateObject();

#if COMPILING_FOR_LUA
luabridge::RefCountedPtr<MyClass> CreateObject_Lua()
{ return CreateObject(); }
#endif
```

If there were a simple way for me to create that object on the Lua stack, does that not seem the best solution? But I don't want to get into a big modification project on every one of those Create functions when `RefCountedPtr` is getting the job done now. A challenge is that I don't trust the copy constructors of these classes, so it would have to be constructed in place.

kunitoki commented:
But your miraculous RefCountedPtr wrapping a shared_ptr, does really nothing more than a shared_ptr, it just add noise and zero benefits for no apparent reason. Give a try to my other PR, should work for your case (just create a template alias to shared_ptr named RefCountedPtr)

kunitoki commented:
There's another option for your script created classes, using LuaBridge3 factories https://kunitoki.github.io/LuaBridge3/Manual#262---constructor-factories

rpatters1 commented:
Your PR is better. As far as I know, my code never tries to get the pointer out back out of Lua. There (probably) are cases where Lua passes the pointer back to C++ through an API, but there is very little interaction with Lua from the C++ side by means of LuaRefs. (Essentially none in the legacy framework itself.)

rpatters1 commented:
I will close this PR.

rpatters1 commented:
I saw factories when I read the docs, but am having trouble seeing how my code would work with a factory. Do factories support null returns from the allocator functions (sending nil to Lua)? Can I make this work while guaranteeing no copy constructors?

If std::shared_ptr will work out of the box, that's what I'm going to use. For cross-DLL compatibility they probably need explicit deallocators, but that's another story.
